File: backend/src/main/java/com/back/ourlog/domain/auth/controller/AuthController.java
package com.back.ourlog.domain.auth.controller;

import com.back.ourlog.domain.auth.dto.LoginRequest;
import com.back.ourlog.domain.auth.dto.LoginResponse;
import com.back.ourlog.domain.auth.dto.OAuthCallbackRequest;
import com.back.ourlog.domain.auth.dto.SignupRequest;
import com.back.ourlog.domain.auth.service.AuthService;
import com.back.ourlog.domain.auth.service.OAuthService;
import com.back.ourlog.global.common.dto.RsData;
import com.back.ourlog.global.security.jwt.TokenDto;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseCookie;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/auth")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;
    private final OAuthService oauthService;

    @Value("${cookie.secure}")
    private boolean secure;

    @PostMapping("/signup")
    public RsData<?> signup(@RequestBody SignupRequest request) {
        authService.signup(request);
        return RsData.success("회원가입 성공");
    }

    @PostMapping("/login")
    public RsData<LoginResponse> login(@RequestBody LoginRequest request,
                                       @RequestHeader(value = "X-Device-Id", required = false) String deviceId,
                                       HttpServletResponse response) {

        var loginResult = authService.login(request.email(), request.password(), deviceId);

        ResponseCookie refreshTokenCookie = ResponseCookie.from("refreshToken", loginResult.refreshToken())
                .httpOnly(true)
                .secure(secure)
                .path("/")
                .maxAge(loginResult.refreshTokenExpiration())
                .sameSite("Strict")
                .build();

        response.addHeader("Set-Cookie", refreshTokenCookie.toString());

        return RsData.success("로그인 성공", new LoginResponse(loginResult.accessToken()));
    }

    @PostMapping("/logout")
    @PreAuthorize("isAuthenticated()")
    public RsData<?> logout(@RequestHeader("Authorization") String authorization,
                            @RequestHeader(value = "X-Device-Id", required = false) String deviceId,
                            HttpServletResponse response) {

        String accessToken = authorization.replace("Bearer ", "");

        authService.logout(accessToken, deviceId);

        ResponseCookie deleteCookie = ResponseCookie.from("refreshToken", "")
                .path("/")
                .maxAge(0)
                .secure(secure)
                .httpOnly(true)
                .build();

        response.addHeader("Set-Cookie", deleteCookie.toString());

        return RsData.success("로그아웃 성공");
    }

    @PostMapping("/reissue")
    public RsData<LoginResponse> reissue(
            @CookieValue(name = "refreshToken", required = false) String refreshToken,
            @RequestHeader(value = "X-Device-Id", required = false) String deviceId,
            HttpServletResponse response) {

        var newTokens = authService.reissue(refreshToken, deviceId);

        ResponseCookie refreshTokenCookie = ResponseCookie.from("refreshToken", newTokens.refreshToken())
                .httpOnly(true)
                .secure(secure)
                .path("/")
                .maxAge(newTokens.refreshTokenExpiration())
                .sameSite("Strict")
                .build();

        response.addHeader("Set-Cookie", refreshTokenCookie.toString());

        return RsData.success("토큰 재발급 성공", new LoginResponse(newTokens.accessToken()));
    }

    @PostMapping("/oauth/callback/{provider}")
    public RsData<LoginResponse> oauthCallback(@PathVariable String provider,
                                               @RequestBody OAuthCallbackRequest request,
                                               @RequestHeader(value = "X-Device-Id", required = false) String deviceId,
                                               HttpServletResponse response) {

        TokenDto tokenDto = oauthService.handleOAuthCallback(provider, request, deviceId);

        ResponseCookie refreshTokenCookie = ResponseCookie.from("refreshToken", tokenDto.refreshToken())
                .httpOnly(true)
                .secure(secure)
                .path("/")
                .maxAge(tokenDto.refreshTokenExpiration())
                .sameSite("Strict")
                .build();

        response.addHeader("Set-Cookie", refreshTokenCookie.toString());

        return RsData.success("OAuth 로그인 성공", new LoginResponse(tokenDto.accessToken()));
    }
}

File: backend/src/main/java/com/back/ourlog/domain/auth/controller/AuthController.kt
package com.back.ourlog.domain.auth.controller

import com.back.ourlog.domain.auth.dto.LoginRequest
import com.back.ourlog.domain.auth.dto.LoginResponse
import com.back.ourlog.domain.auth.dto.OAuthCallbackRequest
import com.back.ourlog.domain.auth.dto.SignupRequest
import com.back.ourlog.domain.auth.service.AuthService
import com.back.ourlog.domain.auth.service.OAuthService
import com.back.ourlog.global.common.dto.RsData
import com.back.ourlog.global.common.extension.toSuccessResponse
import com.back.ourlog.global.common.extension.toSuccessResponseWithoutData
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.ResponseCookie
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/v1/auth")
class AuthController(
    private val authService: AuthService,
    private val oauthService: OAuthService,
    @Value("\${cookie.secure}") private val secure: Boolean
) {

    @PostMapping("/signup")
    fun signup(@RequestBody request: SignupRequest): ResponseEntity<RsData<Nothing>> {
        authService.signup(request)
        return toSuccessResponseWithoutData("회원가입 성공")
    }

    @PostMapping("/login")
    fun login(
        @RequestBody request: LoginRequest,
        @RequestHeader("X-Device-Id") deviceId: String,
        response: HttpServletResponse
    ): ResponseEntity<RsData<LoginResponse>> {
        val tokenDto = authService.login(request.email, request.password, deviceId)
        setRefreshTokenCookie(response, tokenDto.refreshToken, tokenDto.refreshTokenExpiration.seconds)
        return LoginResponse(tokenDto.accessToken).toSuccessResponse("로그인 성공")
    }

    @PostMapping("/logout")
    fun logout(
        @RequestHeader("Authorization") authorization: String,
        @RequestHeader("X-Device-Id") deviceId: String,
        response: HttpServletResponse
    ): ResponseEntity<RsData<Nothing>> {
        val accessToken = authorization.removePrefix("Bearer ")
        authService.logout(accessToken, deviceId)
        deleteRefreshTokenCookie(response)
        return toSuccessResponseWithoutData("로그아웃 성공")
    }

    @PostMapping("/reissue")
    fun reissue(
        @CookieValue("refreshToken") refreshToken: String,
        @RequestHeader("X-Device-Id") deviceId: String,
        response: HttpServletResponse
    ): ResponseEntity<RsData<LoginResponse>> {
        val newTokens = authService.reissue(refreshToken, deviceId)
        setRefreshTokenCookie(response, newTokens.refreshToken, newTokens.refreshTokenExpiration.seconds)
        return LoginResponse(newTokens.accessToken).toSuccessResponse("토큰 재발급 성공")
    }

    @PostMapping("/oauth/callback/{provider}")
    fun oauthCallback(
        @PathVariable provider: String,
        @RequestBody request: OAuthCallbackRequest,
        @RequestHeader("X-Device-Id") deviceId: String,
        response: HttpServletResponse
    ) = kotlinx.coroutines.runBlocking {
        val tokenDto = oauthService.handleOAuthCallback(provider, request, deviceId)
        setRefreshTokenCookie(response, tokenDto.refreshToken, tokenDto.refreshTokenExpiration.seconds)
        LoginResponse(tokenDto.accessToken).toSuccessResponse("OAuth 로그인 성공")
    }

    private fun setRefreshTokenCookie(response: HttpServletResponse, token: String, maxAge: Long) {
        val cookie = ResponseCookie.from("refreshToken", token)
            .httpOnly(true)
            .secure(secure)
            .path("/")
            .maxAge(maxAge)
            .sameSite("Strict")
            .build()
        response.addHeader("Set-Cookie", cookie.toString())
    }

    private fun deleteRefreshTokenCookie(response: HttpServletResponse) {
        val cookie = ResponseCookie.from("refreshToken", "")
            .httpOnly(true)
            .secure(secure)
            .path("/")
            .maxAge(0)
            .sameSite("Strict")
            .build()
        response.addHeader("Set-Cookie", cookie.toString())
    }
}